libs/providers/ofrep/src/lib/ofrep-provider.ts

```diff
@@ -0,0 +1,133 @@
+import {
+  ErrorCode,
+  EvaluationContext,
+  JsonValue,
+  Provider,
+  ResolutionDetails,
+  StandardResolutionReasons,
+  TypeMismatchError,
+} from '@openfeature/server-sdk';
+import {
+  EvaluationFlagValue,
+  handleEvaluationError,
+  HttpHeaders,
+  mergeHeaders,
+  OFREPApi,
+  OFREPApiEvaluationResult,
+  OFREPApiTooManyRequestsError,
+  OFREPProviderBaseOptions,
+  RequestOptions,
+  toRequestOptions,
+  toResolutionDetails,
+} from '@openfeature/ofrep-core';
+
+export type OFREPProviderOptions = Omit<OFREPProviderBaseOptions, 'headersFactory'> & {
+  headersFactory?: () => Promise<HttpHeaders> | HttpHeaders;
+};
+
+export class OFREPProvider implements Provider {
+  private notBefore: Date | null = null;
+  private ofrepApi: OFREPApi;
+
+  readonly runsOn = 'server';
+  readonly metadata = {
+    name: 'OpenFeature Remote Evaluation Protocol Server',
+  };
+
+  constructor(private options: OFREPProviderOptions) {
+    try {
+      // Cannot use URL.canParse as it is only available from Node 19.x
+      new URL(this.options.baseUrl);
+    } catch {
+      throw new Error(`The given OFREP URL "${this.options.baseUrl}" is not a valid URL.`);
+    }
+
+    this.ofrepApi = new OFREPApi(options.baseUrl, options.fetchImplementation);
+  }
+
+  public async resolveBooleanEvaluation(
+    flagKey: string,
+    defaultValue: boolean,
+    context: EvaluationContext,
+  ): Promise<ResolutionDetails<boolean>> {
+    return this.evaluate(flagKey, defaultValue, context);
+  }
+
+  public async resolveStringEvaluation(
+    flagKey: string,
+    defaultValue: string,
+    context: EvaluationContext,
+  ): Promise<ResolutionDetails<string>> {
+    return this.evaluate(flagKey, defaultValue, context);
+  }
+
+  public async resolveNumberEvaluation(
+    flagKey: string,
+    defaultValue: number,
+    context: EvaluationContext,
+  ): Promise<ResolutionDetails<number>> {
+    return this.evaluate(flagKey, defaultValue, context);
+  }
+
+  public async resolveObjectEvaluation<U extends JsonValue>(
+    flagKey: string,
+    defaultValue: U,
+    context: EvaluationContext,
+  ): Promise<ResolutionDetails<U>> {
+    return this.evaluate(flagKey, defaultValue, context);
+  }
+
+  private async evaluate<T extends EvaluationFlagValue>(
+    flagKey: string,
+    defaultValue: T,
+    context: EvaluationContext,
+  ): Promise<ResolutionDetails<T>> {
+    const currentDate = new Date();
+    if (this.notBefore && this.notBefore > currentDate) {
+      return {
+        value: defaultValue,
+        reason: StandardResolutionReasons.DEFAULT,
+        errorCode: ErrorCode.GENERAL,
+        flagMetadata: {
+          retryAfter: this.notBefore.toISOString(),
+        },
+        errorMessage: `OFREP evaluation paused due to TooManyRequests until ${this.notBefore.toISOString()}`,
+      };
+    } else if (this.notBefore) {
+      this.notBefore = null;
+    }
```

Comment:
@beeme1mr @toddbaert @thomaspoignant 
I decided to short circuit evaluations on 429 if there is a value in the Retry-After Header. 
Would you change anything about this resolved value here? 

This is the last thing I would like to have validated with you, afterwards I would merge.

Reply:
We may also throw here to let the SDK take care of the default value + the error code.

Reply:
Yes I thought about it, but then I thought this is a pretty special error that might be up to the provider 🤔

Reply:
I am fine with both, I raised this because in @Kavindu-Dodan go provider an error is sent for this use case, and I would try to stay consistent as much as possible in the different server providers.

https://github.com/open-feature/go-sdk-contrib/blob/ab673a26abaf3fba11c6d28ce1b3ee476c97ab6e/providers/ofrep/internal/evaluate/resolver.go#L68

Reply:
Yes totally and for normal errors I would be totally with you regarding throwing.
Mh I am not sure what we want to do, maybe the others have an opinion?

Reply:
I think we always have to throw right now for errors so the hooks run. I think this is the right thing to do with a 429 for server evals. Effectively this seems to me to be the same as any other error - for instance if the service was down.

So I would change this to throw a General with an informative error.

Reply:
Okay sounds fine to me @toddbaert.
I think I would just propagate the TooManyRequestsError as it will be converted to a GeneralError by the SDK.

Reply:
I went ahead with the error approach as this results in default evaluations. I think end users should act on 429 for example by increasing their evaluation quota at the vendor.

Reply:
I will change it to throwing then :)

Reply:
I changed this now.
@thomaspoignant @beeme1mr @toddbaert @Kavindu-Dodan any obligations of merging it now?

Reply:
For me we are in a good state now.